File: engine/julia/ratio_backspread_engine.jl
module RatioBackspreadEngineModule

export RatioBackspreadState, construct_backspread, calculate_pnl

struct RatioBackspreadState
    short_strike::Float64
    long_strike::Float64
    net_debit_credit::Float64
    max_loss_point::Float64
    upper_bep::Float64
    lower_bep::Float64
    is_call::Bool
end

function construct_backspread(atm::Float64, otm::Float64, short_prem::Float64, long_prem::Float64, is_call::Bool)::RatioBackspreadState
    net_flow = (2.0 * long_prem) - short_prem
    max_loss = abs(otm - atm) + net_flow
    upper_bep = otm + max_loss
    lower_bep = atm + (net_flow < 0.0 ? net_flow : 0.0)
    return RatioBackspreadState(atm, otm, net_flow, max_loss, upper_bep, lower_bep, is_call)
end

function calculate_pnl(state::RatioBackspreadState, st::Float64)::Float64
    if state.is_call
        short_val = max(0.0, st - state.short_strike)
        long_val = 2.0 * max(0.0, st - state.long_strike)
        return (long_val - short_val) - state.net_debit_credit
    else
        short_val = max(0.0, state.short_strike - st)
        long_val = 2.0 * max(0.0, state.long_strike - st)
        return (long_val - short_val) - state.net_debit_credit
    end
end

end

File: engine/rust/src/ratio_backspread_engine.rs
#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct RatioBackspreadState {
    pub short_strike: f64,
    pub long_strike: f64,
    pub net_debit_credit: f64,
    pub max_loss_point: f64,
    pub upper_bep: f64,
    pub lower_bep: f64,
    pub implied_volatility: f64,
    pub ratio_short: u16,
    pub ratio_long: u16,
    pub is_call_spread: u8,
    pub _padding: [u8; 3],
}

impl RatioBackspreadState {
    pub fn new_call_backspread(atm: f64, otm: f64, short_prem: f64, long_prem: f64) -> Self {
        let net_flow = (2.0 * long_prem) - short_prem;
        let max_loss = (otm - atm).abs() + net_flow;
        let upper_bep = otm + max_loss;
        let lower_bep = atm + if net_flow < 0.0 { net_flow } else { 0.0 };

        Self {
            short_strike: atm,
            long_strike: otm,
            net_debit_credit: net_flow,
            max_loss_point: max_loss,
            upper_bep,
            lower_bep,
            implied_volatility: 0.30,
            ratio_short: 1,
            ratio_long: 2,
            is_call_spread: 1,
            _padding: [0; 3],
        }
    }
}
